gordo/machine/dataset/filter_periods.py

```diff
@@ -0,0 +1,205 @@
+    pass
+
+
+class filter_periods:
+    """Model class with methods for data pre-processing.
+    Performs a series of algorithms that drops noisy data.
+
+    Either a rolling median or an isolation forest algorithm is executed.
+    Both provide drop periods in a dict-type element on the class object
+    `object.drop_periods["iforest"]` and `object.drop_periods["median"]`,
+    and data is filtered accordingly.
+
+    Parameters:
+    data: pandas.DataFrame
+        Data frame containing already filtered data (global max/min + dropped known periods).
+        Time consecutively is not required.
+    granularity: str
+        The bucket size for grouping all incoming time data (e.g. "10T").
+        Available strings come from https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html#dateoffset-objects
+    **kwargs:
+        See below.
+
+    Keyword arguments:
+    filter_method: str
+        Which method should be used for data cleaning, either "median" (default), "iforest" or "all" which returns
+        results for both methods.
+    iforest_smooth: bool
+        If exponential weighted smoothing should be applied to data before isolation forest
+        algorithm is run.
+    """
+
+    def __init__(self, data, granularity, **kwargs):
+        self.data = data.copy()
+        self.granularity = granularity
+        self.filter_method = kwargs.get("filter_method", "median")
+
+        if self.filter_method not in ["median", "iforest", "all"]:
+            raise WrongFilterMethodType
+
+        self.predictions = {}
+        if self.filter_method in ["median", "all"]:
+            self._window = kwargs.get("window", 144)
+            self._n_iqr = kwargs.get("n_iqr", 5)
+            self._rolling_median()
+
+        if self.filter_method in ["iforest", "all"]:
+            self._iforest_smooth = kwargs.get("iforest_smooth", False)
+            self._contamination = kwargs.get("contamination", 0.03)
+            self._train()
+            self._predict()
+
+        self._drop_periods()
+        self._filter_data()
+
+    def _init_model(self):
+        """Return a new instance of the models."""
+        self.isolationforest = IsolationForest(
+            n_estimators=300,  # The number of base estimators in the ensemble.
+            max_samples=min(
+                1000, self.data.shape[0]
+            ),  # Samples to draw from X to train each base estimator
+            contamination=self._contamination,  # default = "auto"
+            max_features=1.0,  # Features to draw from X to train each base estimator.
+            bootstrap=False,
+            n_jobs=-1,  # ``-1`` means using all processors
+            random_state=0,
+            verbose=0,
+        )
+        self.minmaxscaler = MinMaxScaler()
+        logger.info("Initialized isolationforest: \n%s.", self.isolationforest)
+        logger.info("Initialized minmaxscaler: \n%s.", self.minmaxscaler)
+
+    def _train(self):
```

Comment:
Passing data through all methods in way of processing chain looks much more readable. And in this case is not required to do `data.copy()` on each of method only if it is really necessary:
```suggestion
    def _train(self, data):
```